k8s/reconcilers/deployment.go

```diff
@@ -0,0 +1,679 @@
+		pollInterval:           defaultPollInterval,
+		deleteTimeOut:          defaultTimeout,
+	}
+	for _, opt := range opts {
+		opt(r)
+	}
+	if r.finalizerName == "" {
+		return nil, fmt.Errorf("finalizer name cannot be empty")
+	}
+	if r.kubeClient == nil {
+		return nil, fmt.Errorf("kube client cannot be nil")
+	}
+	if r.apiClient == nil {
+		return nil, fmt.Errorf("api client cannot be nil")
+	}
+	if r.deploymentBuilder == nil {
+		return nil, fmt.Errorf("deployment builder cannot be nil")
+	}
+	return r, nil
+}
+
+func (r *DeploymentReconciler) deriveReconcileInterval(log logr.Logger, target Reconcilable) (reconciliationInterval, timeout time.Duration) {
+	rp := target.GetReconciliationPolicy()
+	reconciliationInterval = r.reconciliationInterval
+	timeout = r.applyTimeOut
+	if rp != nil && rp.Type == k8smodel.ReconciliationPolicy_Periodic {
+		if rp.Interval != nil {
+			interval, err := time.ParseDuration(*rp.Interval)
+			if err != nil {
+				log.Info(fmt.Sprintf("failed to parse reconciliation interval %s, using default %s", *rp.Interval, reconciliationInterval))
+				return
+			}
+			reconciliationInterval = interval
+		}
+	}
+	return
+}
+
+// attemptUpdate attempts to update the instance
+func (r *DeploymentReconciler) AttemptUpdate(ctx context.Context, object Reconcilable, log logr.Logger, operationStartTimeKey string) (metrics.OperationStatus, reconcile.Result, error) {
+	if !controllerutil.ContainsFinalizer(object, r.finalizerName) {
+		controllerutil.AddFinalizer(object, r.finalizerName)
+		// updates the object in Kubernetes cluster
+		if err := r.kubeClient.Update(ctx, object); err != nil {
+			return metrics.StatusUpdateFailed, ctrl.Result{}, err
+		}
+	}
+
+	if object.GetAnnotations()[operationStartTimeKey] == "" || utilsmodel.IsTerminalState(object.GetStatus().ProvisioningStatus.Status) {
+		r.patchOperationStartTime(object, operationStartTimeKey)
+		if err := r.kubeClient.Update(ctx, object); err != nil {
+			return metrics.StatusUpdateFailed, ctrl.Result{}, err
+		}
+	}
+
+	// Get reconciliation interval
+	reconciliationInterval, timeout := r.deriveReconcileInterval(log, object)
+
+	// If the object hasn't reached a terminal state and the time since the operation started is greater than the
+	// apply timeout, we should update the status with a terminal error and return
+	startTime, err := time.Parse(time.RFC3339, object.GetAnnotations()[operationStartTimeKey])
+	if err != nil {
+		return metrics.StatusUpdateFailed, ctrl.Result{}, err
+	}
+	if time.Since(startTime) > timeout && !utilsmodel.IsTerminalState(object.GetStatus().ProvisioningStatus.Status) {
+		if _, err := r.updateObjectStatus(ctx, object, nil, patchStatusOptions{
+			terminalErr: v1alpha2.NewCOAError(nil, "failed to completely reconcile within the allocated time", v1alpha2.TimedOut),
+		}, log); err != nil {
+			return metrics.StatusUpdateFailed, ctrl.Result{}, err
+		}
+		return metrics.DeploymentTimedOut, ctrl.Result{RequeueAfter: reconciliationInterval}, nil
```

Comment:
Curious what's requeued, another AttemptUpdate action?

Reply:
Requeue tells the controller the requeue the reconcile key after the Duration. You can think in reconciliationInterval, the reconciliation loop will be triggered again. 
We cannot assume it is AttemptUpdate or not. In reconciliation loop, controller will retrieve the object and decide the operation according to its spec.